File: solveZMP.m
function [x_zmp,y_zmp] = solveZMP(x0,y0,z0,x0b,y0b,z0b,x0bb,y0bb,z0bb,...
                                  x1,y1,z1,x1b,y1b,z1b,x1bb,y1bb,z1bb,...
                                  x2,y2,z2,x2b,y2b,z2b,x2bb,y2bb,z2bb,...
                                  x3,y3,z3,x3b,y3b,z3b,x3bb,y3bb,z3bb,...
                                  x4,y4,z4,x4b,y4b,z4b,x4bb,y4bb,z4bb,...
                                  x5,y5,z5,x5b,y5b,z5b,x5bb,y5bb,z5bb,...
                                  x6,y6,z6,x6b,y6b,z6b,x6bb,y6bb,z6bb,...
                                  m0,m1,m2,m3,m4,m5,m6,g,N,tf,ts)
    tsample=(tf-ts)/N;
    ax0=(x0+x0bb-2*x0b)/(tsample^2);
    ax1=(x1+x1bb-2*x1b)/(tsample^2);
    ax2=(x2+x2bb-2*x2b)/(tsample^2);
    ax3=(x3+x3bb-2*x3b)/(tsample^2);
    ax4=(x4+x4bb-2*x4b)/(tsample^2);
    ax5=(x5+x5bb-2*x5b)/(tsample^2);
    ax6=(x6+x6bb-2*x6b)/(tsample^2);
    ay0=(y0+y0bb-2*y0b)/(tsample^2);
    ay1=(y1+y1bb-2*y1b)/(tsample^2);
    ay2=(y2+y2bb-2*y2b)/(tsample^2);
    ay3=(y3+y3bb-2*y3b)/(tsample^2);
    ay4=(y4+y4bb-2*y4b)/(tsample^2);
    ay5=(y5+y5bb-2*y5b)/(tsample^2);
    ay6=(y6+y6bb-2*y6b)/(tsample^2);
    az0=(z0+z0bb-2*z0b)/(tsample^2);
    az1=(z1+z1bb-2*z1b)/(tsample^2);
    az2=(z2+z2bb-2*z2b)/(tsample^2);
    az3=(z3+z3bb-2*z3b)/(tsample^2);
    az4=(z4+z4bb-2*z4b)/(tsample^2);
    az5=(z5+z5bb-2*z5b)/(tsample^2);
    az6=(z6+z6bb-2*z6b)/(tsample^2);
    
    x_zmp=(m0*(az0*x0+g*x0-ax0*z0)+m1*(az1*x1+g*x1-ax1*z1)+...
           m2*(az2*x2+g*x2-ax2*z2)+m3*(az3*x3+g*x3-ax3*z3)+...
           m4*(az4*x4+g*x4-ax4*z4)+m5*(az5*x5+g*x5-ax5*z5)+...
           m6*(az6*x6+g*x6-ax6*z6))/(m0*(az0+g)+m1*(az1+g)+...
           m2*(az2+g)+m3*(az3+g)+m4*(az4+g)+m5*(az5+g)+...
           m6*(az6+g));
    y_zmp=(m0*(az0*y0+g*y0-ay0*z0)+m1*(az1*y1+g*y1-ay1*z1)+...
           m2*(az2*y2+g*y2-ay2*z2)+m3*(az3*y3+g*y3-ay3*z3)+...
           m4*(az4*y4+g*y4-ay4*z4)+m5*(az5*y5+g*y5-ay5*z5)+...
           m6*(az6*y6+g*y6-ay6*z6))/(m0*(az0+g)+m1*(az1+g)+...
           m2*(az2+g)+m3*(az3+g)+m4*(az4+g)+m5*(az5+g)+...
           m6*(az6+g));
end

File: solveZMP.py
import math
import numpy
import matplotlib.pyplot as plt
from parameter import t
from parameter import x_f
from parameter import z_f
from parameter import x_h
from parameter import z_h
from parameter import x_m
from parameter import z_m
from parameter import x_l
from parameter import z_l
from parameter import x_m2
from parameter import z_m2
from parameter import size

def list_add_d(a,b):
    c = []
    for i in range(len(a)):
        c.append((a[i]+b[i])/2)
    return c

def funax(a,t_sample):
    res = []
    for i in range(len(a) - 2):
        res.append((a[i + 2] + a[i] - 2*a[i + 1])/ t_sample**2)
    res.append(res[len(res) - 1])
    res.append(res[len(res) - 1])
    return res

def funzmp(x,z,ax,az,m,g):
    mol,den = 0,0
    for i in range(len(x)):
        mol += m[i]*((az[i]+g)*x[i] - ax[i]*z[i])
        den += m[i]*(az[i]+g)
    return mol / den

t_sample = 0.001
g = 9.81
m1,m2,m3,m4,m5,m6,m7=1.52,1.68,2.68,10.45,2.68,1.68,1.52
x1 = x_l
z1 = []
for i in range(size):
    z1.append(-0.04)
x2 = list_add_d(x_l,x_m2)
z2 = list_add_d(z_l,z_m2)
x3 = list_add_d(x_m2,x_h)
z3 = list_add_d(z_m2,z_h)
x4 = x_h
z4 = []
for i in range(size):
    z4.append(z_h[i] + 0.12)
x5 = list_add_d(x_h,x_m)
z5 = list_add_d(z_h,z_m)
x6 = list_add_d(x_m,x_f)
z6 = list_add_d(z_m,z_f)
x7 = x_f
z7 = []
for i in range(size):
    z7.append(z_f[i] - 0.04)
ax1 = funax(x1,t_sample)
ax2 = funax(x2,t_sample)
ax3 = funax(x3,t_sample)
ax4 = funax(x4,t_sample)
ax5 = funax(x5,t_sample)
ax6 = funax(x6,t_sample)
ax7 = funax(x7,t_sample)
az1 = funax(z1,t_sample)
az2 = funax(z2,t_sample)
az3 = funax(z3,t_sample)
az4 = funax(z4,t_sample)
az5 = funax(z5,t_sample)
az6 = funax(z6,t_sample)
az7 = funax(z7,t_sample)

x_zmp = []

for i in range(size):
    x = [x1[i],x2[i],x3[i],x4[i],x5[i],x6[i],x7[i]]
    z = [z1[i],z2[i],z3[i],z4[i],z5[i],z6[i],z7[i]]
    ax = [ax1[i],ax2[i],ax3[i],ax4[i],ax5[i],ax6[i],ax7[i]]
    az = [az1[i],az2[i],az3[i],az4[i],az5[i],az6[i],az7[i]]
    m = [m1,m2,m3,m4,m5,m6,m7]
    tmp = funzmp(x,z,ax,az,m,g)
    x_zmp.append(tmp)

modzmp = 0
for i in range(size):
    modzmp += (abs(x_zmp[i]) > 0.12)

fig = plt.figure(1)
ax = fig.add_subplot(1,1,1)
ax.plot(t,x_zmp,color = 'r')
plt.show()